src/core/set.c

```diff
@@ -198,58 +315,212 @@ igraph_integer_t igraph_set_size(const igraph_set_t* set) {
  * Time complexity: O(log(n)), n is the number of elements in \p set.
  */
 igraph_error_t igraph_set_add(igraph_set_t* set, igraph_integer_t e) {
-    igraph_integer_t left, right, middle;
-    igraph_integer_t size;
-    IGRAPH_ASSERT(set != NULL);
-    IGRAPH_ASSERT(set->stor_begin != NULL);
-
-    size = igraph_set_size(set);
-
-    /* search where to insert the new element */
-    left = 0;
-    right = size - 1;
-    while (left < right - 1) {
-        middle = (left + right) / 2;
-        if (SET(*set)[middle] > e) {
-            right = middle;
-        } else if (SET(*set)[middle] < e) {
-            left = middle;
-        } else {
-            left = middle;
-            break;
-        }
+    if (igraph_set_contains(set, e)) {
+        return IGRAPH_SUCCESS;
     }
 
-    if (right >= 0 && SET(*set)[left] != e && SET(*set)[right] == e) {
-        left = right;
+    struct Node* newNode = IGRAPH_CALLOC(1, struct Node);
```

Comment:
If every `set_add()` involves a memory allocation, that is going be problematic for performance, as `malloc()` is expensive.